### PR TITLE
Top level data source templates

### PR DESCRIPTION
**Pull request by leoyvens**
See https://github.com/graphprotocol/graph-node/issues/1023. The old format is immediately unsupported, but we print a helpful error message.

One change that was not in the plan is the path to the codegen'ed files. The template classes now all go in `generated/templates.ts` and their ABIs in `generated/templates/<template_name>/<abi_name>.ts`.

This also does a migration to bump the spec version to 0.0.2, since this is a breaking change.

### CODE REVIEW

File: src/validation/manifest.js [hunk 1]
```diff
@@ -111,16 +111,23 @@ const validators = immutable.fromJS({
...
-                    immutable.fromJS({
-                      path: ctx.get('path'),
-                      message: `Unexpected key in map: ${key}`,
-                    })
+                    key == 'templates'
+                      ? immutable.fromJS({
+                          path: ctx.get('path'),
+                          message: `The way to declare data source templates has changed, ` + 
+                                `please move the templates from inside data sources to ` +
+                                `a \`templates:\` field at the top-level of the manifest.`,
```
Comment (Jannis) [inline src/validation/manifest.js]: `top-level` -> `top level`. `top level` is the noun, `top-level` is more like an adjective, I think.

File: tests/cli/validation.test.js [hunk 2]
```diff
@@ -128,4 +128,12 @@ describe('Validation', () => {
       exitCode: 1,
     },
   )
+  cliTest(
+    'deprecated template format gives nice error',
```
Comment (Jannis) [inline tests/cli/validation.test.js]: Nit: all test names start with a capital letter so far.

**Review by Jannis — CHANGES_REQUESTED** [review]

**Comment by leoyvens** [comment]
@Jannis comments addressed.

File: src/type-generator.js [hunk 3]
```diff
@@ -274,94 +272,67 @@ module.exports = class TypeGenerator {
...
+            step(
+              spinner,
+              'Generate types for data source template',
+              `${template.get('name')}`,
+            )
+
+            let codeGenerator = new DataSourceTemplateCodeGenerator(template)
+
+            // Only generate module imports once, because they are identical for all
+            // types generate for data source templates
```
Comment (fordN) [inline src/type-generator.js]: generated?

**Review by fordN — APPROVED** [review]
Looks good! I just noticed a very minor tense typo, see my comment.

File: src/type-generator.js (hunk 3 above)
Comment (leoyvens) [reply src/type-generator.js]: Fixed, thanks.

**Review by Jannis — APPROVED** [review]

**Comment by leoyvens** [comment]
Rebased, and fixed a bug with the output directory of template mappings. @Jannis could you review again?

**Pull request merged**